In [10]:
import pandas as pd
import numpy as np
%matplotlib inline
from matplotlib import pyplot as plt
import matplotlib
matplotlib.rcParams["figure.figsize"] = (20, 10)

In [12]:
df1 = pd.read_csv('Bengaluru_House_Data.csv')
df1.head()

area_type   availability                  location       size  \
0  Super built-up  Area         19-Dec  Electronic City Phase II      2 BHK   
1            Plot  Area  Ready To Move          Chikka Tirupathi  4 Bedroom   
2        Built-up  Area  Ready To Move               Uttarahalli      3 BHK   
3  Super built-up  Area  Ready To Move        Lingadheeranahalli      3 BHK   
4  Super built-up  Area  Ready To Move                  Kothanur      2 BHK   

   society total_sqft  bath  balcony   price  
0  Coomee        1056   2.0      1.0   39.07  
1  Theanmp       2600   5.0      3.0  120.00  
2      NaN       1440   2.0      3.0   62.00  
3  Soiewre       1521   3.0      1.0   95.00  
4      NaN       1200   2.0      1.0   51.00

In [13]:
df1.shape

(13320, 9)

In [27]:
df1.groupby('area_type')['area_type'].agg('count').sort_values(ascending=False)

area_type
Super built-up  Area    8790
Built-up  Area          2418
Plot  Area              2025
Carpet  Area              87
Name: area_type, dtype: int64

# 1) Data cleaning

In [28]:
# we assume that some columns are not important to determine price
df2 = df1.drop(['area_type','availability','society','balcony'],axis='columns')
df2.head()

location       size total_sqft  bath   price
0  Electronic City Phase II      2 BHK       1056   2.0   39.07
1          Chikka Tirupathi  4 Bedroom       2600   5.0  120.00
2               Uttarahalli      3 BHK       1440   2.0   62.00
3        Lingadheeranahalli      3 BHK       1521   3.0   95.00
4                  Kothanur      2 BHK       1200   2.0   51.00

In [36]:
# null values 
df2.isnull().sum().sort_values()

total_sqft     0
price          0
location       1
size          16
bath          73
dtype: int64

In [43]:
# we have small number of null values in each column so we can drop corresponding rows
df3 = df2.dropna(axis=0)
df3.isna().sum()

location      0
size          0
total_sqft    0
bath          0
price         0
dtype: int64

In [44]:
df3.shape

(13246, 5)

In [46]:
# size column
df3['size'].unique()

array(['2 BHK', '4 Bedroom', '3 BHK', '4 BHK', '6 Bedroom', '3 Bedroom',
       '1 BHK', '1 RK', '1 Bedroom', '8 Bedroom', '2 Bedroom',
       '7 Bedroom', '5 BHK', '7 BHK', '6 BHK', '5 Bedroom', '11 BHK',
       '9 BHK', '9 Bedroom', '27 BHK', '10 Bedroom', '11 Bedroom',
       '10 BHK', '19 BHK', '16 BHK', '43 Bedroom', '14 BHK', '8 BHK',
       '12 Bedroom', '13 BHK', '18 Bedroom'], dtype=object)

In [55]:
# multiple ways to write number of beedroom --> we need a clean 
df3['bhk'] = df3['size'].apply(lambda x : int(x.split(" ")[0]))

/tmp/ipykernel_32076/1328507227.py:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df3['bhk'] = df3['size'].apply(lambda x : int(x.split(" ")[0]))


In [56]:
df3.head()

location       size total_sqft  bath   price  bhk
0  Electronic City Phase II      2 BHK       1056   2.0   39.07    2
1          Chikka Tirupathi  4 Bedroom       2600   5.0  120.00    4
2               Uttarahalli      3 BHK       1440   2.0   62.00    3
3        Lingadheeranahalli      3 BHK       1521   3.0   95.00    3
4                  Kothanur      2 BHK       1200   2.0   51.00    2

In [71]:
bhk = list(df3['bhk'].unique())
bhk.sort()
bhk

[1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 16, 18, 19, 27, 43]

In [76]:
df3[df3['bhk']>20]

location        size total_sqft  bath  price  bhk
1718  2Electronic City Phase II      27 BHK       8000  27.0  230.0   27
4684                Munnekollal  43 Bedroom       2400  40.0  660.0   43

In [78]:
# we have a house that contains 43 bedroon but with 2400 total_sqft : looks like an error

# analyse total_sqft column
df3.total_sqft.unique()

array(['1056', '2600', '1440', ..., '1133 - 1384', '774', '4689'],
      dtype=object)

In [81]:
def is_float(x):
    try: 
        float(x)
    except:
        return False
    else:
        return True
    

In [87]:
df3[~df3['total_sqft'].apply(is_float)].head(20)

location       size      total_sqft  bath    price  bhk
30                Yelahanka      4 BHK     2100 - 2850   4.0  186.000    4
122                  Hebbal      4 BHK     3067 - 8156   4.0  477.000    4
137      8th Phase JP Nagar      2 BHK     1042 - 1105   2.0   54.005    2
165                Sarjapur      2 BHK     1145 - 1340   2.0   43.490    2
188                KR Puram      2 BHK     1015 - 1540   2.0   56.800    2
410                 Kengeri      1 BHK  34.46Sq. Meter   1.0   18.500    1
549             Hennur Road      2 BHK     1195 - 1440   2.0   63.770    2
648                 Arekere  9 Bedroom       4125Perch   9.0  265.000    9
661               Yelahanka      2 BHK     1120 - 1145   2.0   48.130    2
672            Bettahalsoor  4 Bedroom     3090 - 5002   4.0  445.000    4
772   Banashankari Stage VI      2 BHK     1160 - 1195   2.0   59.935    2
775            Basavanagara      1 BHK   1000Sq. Meter   2.0   93.000    1
850       Bannerghatta Road      2 BHK     1115 - 1130   2.0   58.935    2
872       Singapura Village      2 BHK   1100Sq. Yards   2.0   45.000    2
886              Chandapura      1 BHK       520 - 645   1.0   15.135    1
927             Thanisandra      2 BHK     1000 - 1285   2.0   43.415    2
959             Kammasandra      1 BHK       650 - 665   1.0   18.410    1
990                Sarjapur      1 BHK       633 - 666   1.0   17.535    1
1019         Marathi Layout  1 Bedroom       5.31Acres   1.0  110.000    1
1086             Narasapura  2 Bedroom         30Acres   2.0   29.500    2

In [96]:
# there are different ways to write total_sqft --> we need a clean

def convert_sqft_to_num(x):
    tokens = x.split('-')
    if len(tokens) == 2:
        return ((float(tokens[0].strip())+ float(tokens[1].strip()))/2 )
    try:
        return float(x)
    except:
        return None

In [97]:
df4 = df3.copy()
df4['total_sqft'] = df4['total_sqft'].apply(convert_sqft_to_num)

In [98]:
df4

location       size  total_sqft  bath   price  bhk
0      Electronic City Phase II      2 BHK      1056.0   2.0   39.07    2
1              Chikka Tirupathi  4 Bedroom      2600.0   5.0  120.00    4
2                   Uttarahalli      3 BHK      1440.0   2.0   62.00    3
3            Lingadheeranahalli      3 BHK      1521.0   3.0   95.00    3
4                      Kothanur      2 BHK      1200.0   2.0   51.00    2
...                         ...        ...         ...   ...     ...  ...
13315                Whitefield  5 Bedroom      3453.0   4.0  231.00    5
13316             Richards Town      4 BHK      3600.0   5.0  400.00    4
13317     Raja Rajeshwari Nagar      2 BHK      1141.0   2.0   60.00    2
13318           Padmanabhanagar      4 BHK      4689.0   4.0  488.00    4
13319              Doddathoguru      1 BHK       550.0   1.0   17.00    1

[13246 rows x 6 columns]

In [101]:
df4.isna().sum()

location       0
size           0
total_sqft    46
bath           0
price          0
bhk            0
dtype: int64

# 2) Feature engineering 

In [108]:
df5 = df4.copy()

df5['price_per_sqft'] = df5['price']*10000/df5['total_sqft']
df5.head()

location       size  total_sqft  bath   price  bhk  \
0  Electronic City Phase II      2 BHK      1056.0   2.0   39.07    2   
1          Chikka Tirupathi  4 Bedroom      2600.0   5.0  120.00    4   
2               Uttarahalli      3 BHK      1440.0   2.0   62.00    3   
3        Lingadheeranahalli      3 BHK      1521.0   3.0   95.00    3   
4                  Kothanur      2 BHK      1200.0   2.0   51.00    2   

   price_per_sqft  
0      369.981061  
1      461.538462  
2      430.555556  
3      624.589086  
4      425.000000

In [110]:
len(df5.location.unique())

1304

In [111]:
df5.location.value_counts()

Whitefield           534
Sarjapur  Road       392
Electronic City      302
Kanakpura Road       266
Thanisandra          233
                    ... 
Vidyapeeta             1
Maruthi Extension      1
Okalipura              1
Old Town               1
Abshot Layout          1
Name: location, Length: 1304, dtype: int64

### location is a categorical column  : we have 1304 value --> we need to reduce them to improve our data quality



In [112]:
df5.location = df5.location.apply(lambda x: x.strip())

In [113]:
len(df5.location.unique())

1293

In [116]:
pd.set_option('display.max_rows',2000)
locations_stats = df5.location.value_counts()

In [123]:
len(locations_stats[locations_stats<=10])

1052

In [129]:
 locations_other = list(locations_stats[locations_stats<=10].index)

In [132]:
df5.location = df5.location.apply(lambda x : 'other' if x in locations_other else x)

In [134]:
len(df5.location.unique())

242

#  3) outiler removal

In [135]:
df5

location       size  total_sqft  bath   price  bhk  \
0      Electronic City Phase II      2 BHK      1056.0   2.0   39.07    2   
1              Chikka Tirupathi  4 Bedroom      2600.0   5.0  120.00    4   
2                   Uttarahalli      3 BHK      1440.0   2.0   62.00    3   
3            Lingadheeranahalli      3 BHK      1521.0   3.0   95.00    3   
4                      Kothanur      2 BHK      1200.0   2.0   51.00    2   
...                         ...        ...         ...   ...     ...  ...   
13315                Whitefield  5 Bedroom      3453.0   4.0  231.00    5   
13316                     other      4 BHK      3600.0   5.0  400.00    4   
13317     Raja Rajeshwari Nagar      2 BHK      1141.0   2.0   60.00    2   
13318           Padmanabhanagar      4 BHK      4689.0   4.0  488.00    4   
13319              Doddathoguru      1 BHK       550.0   1.0   17.00    1   

       price_per_sqft  
0          369.981061  
1          461.538462  
2          430.555556  
3          624.589086  
4          425.000000  
...               ...  
13315      668.983493  
13316     1111.111111  
13317      525.854514  
13318     1040.733632  
13319      309.090909  

[13246 rows x 7 columns]

In [138]:
# the typical sqft per bedroom is 300 --> we gonna use this to identify outliers

df5[df5['total_sqft']/df5['bhk']<300].head(10)

location       size  total_sqft  bath  price  bhk  \
9                  other  6 Bedroom      1020.0   6.0  370.0    6   
45            HSR Layout  8 Bedroom       600.0   9.0  200.0    8   
58         Murugeshpalya  6 Bedroom      1407.0   4.0  150.0    6   
68   Devarachikkanahalli  8 Bedroom      1350.0   7.0   85.0    8   
70                 other  3 Bedroom       500.0   3.0  100.0    3   
78      Kaval Byrasandra      2 BHK       460.0   1.0   22.0    2   
89          Rajaji Nagar  6 Bedroom       710.0   6.0  160.0    6   
119          Hennur Road  2 Bedroom       276.0   3.0   23.0    2   
129   Vishwapriya Layout  7 Bedroom       950.0   7.0  115.0    7   
149                other  6 Bedroom      1034.0   5.0  185.0    6   

     price_per_sqft  
9       3627.450980  
45      3333.333333  
58      1066.098081  
68       629.629630  
70      2000.000000  
78       478.260870  
89      2253.521127  
119      833.333333  
129     1210.526316  
149     1789.168279

In [140]:
df6 = df5[~(df5['total_sqft']/df5['bhk']<300)]

In [142]:
df6.head()

location       size  total_sqft  bath   price  bhk  \
0  Electronic City Phase II      2 BHK      1056.0   2.0   39.07    2   
1          Chikka Tirupathi  4 Bedroom      2600.0   5.0  120.00    4   
2               Uttarahalli      3 BHK      1440.0   2.0   62.00    3   
3        Lingadheeranahalli      3 BHK      1521.0   3.0   95.00    3   
4                  Kothanur      2 BHK      1200.0   2.0   51.00    2   

   price_per_sqft  
0      369.981061  
1      461.538462  
2      430.555556  
3      624.589086  
4      425.000000

In [143]:
df6.shape

(12502, 7)

In [144]:
df6.price_per_sqft.describe()

count    12456.000000
mean       630.850283
std        416.812734
min         26.782981
25%        421.052632
50%        529.411765
75%        691.666667
max      17647.058824
Name: price_per_sqft, dtype: float64

In [151]:
def remove_pps_outliers(df):
    df_out = pd.DataFrame()
    for key, sub_def in df.groupby('location'):
        m = np.mean(sub_def['price_per_sqft'])
        st = np.std(sub_def['price_per_sqft'])
        reduced_df = sub_def[(sub_def['price_per_sqft']>(m-st)) & (sub_def['price_per_sqft']<= (m+st)) ]
        df_out = pd.concat([df_out, reduced_df], ignore_index=True)
    return df_out
        

In [153]:
df7 = remove_pps_outliers(df6)
df7.head()

location   size  total_sqft  bath  price  bhk  price_per_sqft
0  1st Block Jayanagar  4 BHK      2850.0   4.0  428.0    4     1501.754386
1  1st Block Jayanagar  3 BHK      1630.0   3.0  194.0    3     1190.184049
2  1st Block Jayanagar  3 BHK      1875.0   2.0  235.0    3     1253.333333
3  1st Block Jayanagar  3 BHK      1200.0   2.0  130.0    3     1083.333333
4  1st Block Jayanagar  2 BHK      1235.0   2.0  148.0    2     1198.380567

In [154]:
df7.shape

(10241, 7)